Hot-Rod-Transactions-NonXA.asciidoc

```diff
@@ -0,0 +1,313 @@
+The following information is sent:
+
+* XID - this is the transaction id;
+* For each key, it sends:
+** The key;
+** The `ReadState`;
+** The version read (if any);
+** The new value;
+** The metadata;
+
+[NOTE]
+======
+The `ReadState` represents the following information:
+
+* `NOT_READ`: the key wasn't read before the write;
+* `NON_EXISTING`: the key was read before write and it didn't exist;
+* `READ`: The key was read before write and it exists (the versionr read is required)
+======
+
+In the server side, when the request is received, it replays the transaction.
+A new transaction is created and set its XID based on the XID received (to avoid any XID translation).
+
+For each key received, the server performs the following:
+
+* It checks the `ReadState` value:
+** `NOT_READ`: it doesn't fetch the previous value.
+** `NON_EXISTING`: it checks if the key doesn't exist.
+If the key exists, the transaction is rolled-back.
+** `READ`: it fetches the value and metadata.
+Then it compares the version received with the version stored in the metadata.
+If they don't match, the transaction is rolled-back.
+* Finally, it performs the write.
+
+When all keys are handled, the server invokes `EmbeddedTransaction.runPrepare()`.
+The outcome is sending back to the client.
+
+The client receives the outcome and marks the transaction to rollback if the server failed to prepare.
+
+Finally, in `afterCompletion()`, the client only sends the commit or rollback request:
+
+* XID
+* `COMMIT` or `ROLLBACK` request.
+
+In the server, if the transaction was executed locally, it invokes the `EmbeddedTransaction.runCommit(`).
+Otherwise, use the `RecoveryManager` to commit or rollback the transaction.
+
+NOTE: Not sure if the `RecoveryManager` can be used or if I need to create a Global Transaction Table.
+
+== Hot Rod Protocol Details
+
+=== Operation Codes
+
+4 new operation codes:
+
+|===
+|Operation Code | Response Operation Code |Operation
+
+|0x3B
+|0x3C
+|Prepare Request
+
+|0x3D
+|0x3E
+|Commit Request
+
+|0x3F
+|0x40
+|Rollback Request
+
+|===
+
+=== Messages description
+
+The transaction id (XID) is present on the request header.
+However, it will be changed to support variable size XID
+
+.XID
+[cols="<.^,<.^,<"]
+|===
+|Field Name |Size |Value
+
+|Format ID
+|long
+| The XID format.
+
+|Length of Global Transaction id
+|byte
+|The length of global transaction id byte array. It max value is `64`.
+
+|Global Transaction Id
+|byte array
+|The global transaction id.
+
+|Length of Branch Qualifier
+|byte
+|The length of branch qualifier byte array. It max value is `64`.
+
+|Branch Qualifier
+|byte array
+|The branch qualifier.
+
+|===
+
+
+.Prepare Request (operation code `0x3B`)
+[cols="<.^,<.^,<"]
+|===
+|Field Name |Size |Value
+
+|Header
+|variable
+|Request header
+
+|OnePhaseCommit
+|byte
+|When it is set to `1`, the server will use one-phase-commit if available (XA only)
+
+|Number of keys
+|vInt
+|The number of keys
+
+3+^|For each key
```

Comment:
I'm not sure I understand why we have to send all of this for each key involved in tx. I mean, a client has executed a put/remove...etc within tx and each of those have a message id. Couldn't you just send those message ids in the prepare function? It would require the server to track those messages but it would make the prepare lighter? Just wondering :)